### ISSUE TITLE
[HOLD App # 28618] Request & send money - User is redirected to Amount page from Amount page from deep link

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#29972]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.3.87-1

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**If this was caught during regression testing, add the test name, ID and link from TestRail:**

**Email or phone of affected tester (no customers):**

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Expensify/Expensify Issue URL:**

**Issue reported by:** Applause - Internal Team

**Slack conversation:**

**Issue found when executing PR** https://github.com/Expensify/App/pull/29655

## Action Performed:
1. Go to staging.new.expensify.com
2. Paste Request money link (https://staging.new.expensify.com/request/new
) to any chat,
3, Click on the link above.
4. Enter amount and click Next.
5. Repeat Step 2-4 with Send money link (https://staging.new.expensify.com/send/new
).

## Expected Result:
User will be redirected to the next step of the Request money/Send money process

## Actual Result:
The process is looped. Instead of proceeding to the next step, user lands on the enter amount page again

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

 

https://github.com/Expensify/App/assets/78819774/39517ebe-aec8-4659-b056-7511003924ce

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by FitseTLT** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
User redirected from amount page to amount page (instead of participants page) when using deep link
### What is the root cause of that problem?
The problem occurs when there is a mismatch between `iouType` (which is found from route params) and `iou.id` which is found from the `IOU` value saved in `ONYX`. So if the value of the iou was 'request' and we are creating 'send' iou or vice versa `navigateToNextPage` resets the iou value. Because of this line 
https://github.com/Expensify/App/blob/eb5fec2c258e74a69bbed9d93100db4b30b265ec/src/libs/actions/IOU.js#L2863-L2871
But the `resetMoneyRequestInfo` function resets all the iou properties including `iou.amount` (to 0). So in the `MoneyRequestParticipantsPage` this lines of code will navigate it back because `iou.amount` is 0.
https://github.com/Expensify/App/blob/eb5fec2c258e74a69bbed9d93100db4b30b265ec/src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsPage.js#L97-L99
### What changes do you think we should make in order to solve the problem?

We need to update `resetMoneyRequestInfo` to have the option not to reset the amount if needed. Like this,

`function resetMoneyRequestInfo(id = '', donResetAmount = false) {
    const created = currentDate || format(new Date(), CONST.DATE.FNS_FORMAT_STRING);
    Onyx.merge(ONYXKEYS.IOU, {
        id,
        ...(!donResetAmount && {amount: 0}),
        currency: lodashGet(currentUserPersonalDetails, 'localCurrencyCode', CONST.CURRENCY.USD),
        comment: '',
        participants: [],
        merchant: CONST.TRANSACTION.DEFAULT_MERCHANT,
        category: '',
        tag: '',
        created,
        receiptPath: '',
        receiptFilename: '',
        transactionID: '',
        billable: null,
    });
}`
and  replace 
https://github.com/Expensify/App/blob/eb5fec2c258e74a69bbed9d93100db4b30b265ec/src/libs/actions/IOU.js#L2863-L2871

to 

`
function navigateToNextPage(iou, iouType, report, path = '') {
    const moneyRequestID = `${iouType}${report.reportID || ''}`;
    const shouldReset = iou.id !== moneyRequestID;

    // If the money request ID in Onyx does not match the ID from params, we want to start a new request
    // with the ID from params. We need to clear the participants in case the new request is initiated from FAB.
    if (shouldReset) {
        resetMoneyRequestInfo(moneyRequestID, true);
    }
`
Result

https://github.com/Expensify/App/assets/38649957/6bfcf33e-35a8-4c27-b10c-c7e1d6de9689


### What alternative solutions did you explore? (Optional)

**Comment by cooldev900** [comment]
## Proposal
from: @cooldev900
### Please re-state the problem that we are trying to solve in this issue.
If a user clicks on a link to request money (eg: https://staging.new.expensify.com/request/new), enter an amount, click next and then proceed to click on a link to send money (eg: https://staging.new.expensify.com/send/new), the amount will be pre-populated but after clicking next, it will render the amount screen again.
### What is the root cause of that problem?
The problem is that a single IOU object is used for both send and request money screens, and the id of IOU object will become `""` after new request is handled because both sendMoney(send) and createTransaction(request) in MoneyRequestConfirmPage calls resetMoneyRequestInfo function without id.
https://github.com/Expensify/App/blob/8dba113c233c0a433e81a637ea6c28ca815b3f02/src/libs/actions/IOU.js#L124-L141

https://github.com/Expensify/App/blob/302cebf14c4e0fcd6570ac66e9d85a1ca0212ae6/src/pages/iou/steps/MoneyRequestConfirmPage.js#L295-L311

https://github.com/Expensify/App/blob/302cebf14c4e0fcd6570ac66e9d85a1ca0212ae6/src/pages/iou/steps/MoneyRequestConfirmPage.js#L295-L311

https://github.com/Expensify/App/blob/8dba113c233c0a433e81a637ea6c28ca815b3f02/src/libs/actions/IOU.js#L1259-L1279

https://github.com/Expensify/App/blob/8dba113c233c0a433e81a637ea6c28ca815b3f02/src/libs/actions/IOU.js#L2513-L2519
![image](https://github.com/Expensify/App/assets/13826499/ded07f69-e67f-456c-b2df-d26395966867)


After clicking request button in MoneyRequestConfirmPage, the id of IOU object will be empty so shouldReset becomes true due to ` "" !== "request" or "send"`  and then `!isDistanceRequest && ((iou.amount === 0 && !iou.receiptPath) || shouldReset)` turns to true so that navigateBack function lets the user to get back to amount page again.
And 
```
if (shouldReset) {
      IOU.resetMoneyRequestInfo(moneyRequestId);
  }
```
could be another reason in next refresh.

https://github.com/Expensify/App/blob/302cebf14c4e0fcd6570ac66e9d85a1ca0212ae6/src/pages/iou/steps/MoneyRequstParticipantsPage/MoneyRequestParticipantsPage.js#L92-L99

### What changes do you think we should make in order to solve the problem?

Calling the resetMoneyRequestInfo in IOU.js let a user get back to report page automatically but when we visit request/new page directly, we need to take an account if the iou.id is set or not to avoid getting back to amount page again.
```
useEffect(() => {
        if (!iou.id && (lodashGet(route, 'params.iouType', '') === CONST.IOU.TYPE.REQUEST || lodashGet(route, 'params.iouType', '') === CONST.IOU.TYPE.SEND)) {
            IOU.resetMoneyRequestInfo(lodashGet(route, 'params.iouType', ''));
        }
    }, [iou.id, route.params])
```
### What alternative solutions did you explore? (Optional)
N/A

**Comment by maxconnectAbhi** [comment]
# Proposal

## Please re-state the problem that we are trying to solve in this issue
If a user clicks on a link to request money (eg: https://staging.new.expensify.com/request/new), enter an amount, click next and then proceed to click on a link to send money (eg: https://staging.new.expensify.com/send/new), the amount will be pre-populated but after clicking next, it will render the amount screen again.

## What is the root cause of that problem?
The issue here is that the same IOU object 
```
{
amount:0
category:""
comment:""
created:"2023-10-19"
currency:"INR"
id:"request"
merchant:"Request"
participants:[]
receiptFilename:""
receiptPath:""
tag:""
transactionID:"8311076577222983710"
}
```
is used for both sending and requesting money but with opposite id's i.e, When you send money, it's labeled with the `"receive"` ID, and when you request money, it's labeled `"send"`.

Which results : iou.id = 'send' and iouType = 'request'.
This is handled in navigateToNextPage function 
https://github.com/Expensify/App/blob/18acd82f40153fc1e5091aea3217c532142e500e/src/libs/actions/IOU.js#L2863-L2871
which creates the new request

## What changes do you think we should make in order to solve the problem?
We need to reset the IOU if id is not matched through `resetMoneyRequestInfo(moneyRequestID) ` in else case here

https://github.com/Expensify/App/blob/18acd82f40153fc1e5091aea3217c532142e500e/src/pages/iou/steps/NewRequestAmountPage.js#L96-L121

like 
```
else{
            const moneyRequestID = `${iouType}${reportID}`;
            const shouldReset = iou.id !== moneyRequestID;
            if (shouldReset) {
                IOU.resetMoneyRequestInfo(moneyRequestID);
            }
```
## What alternative solutions did you explore? (Optional)
NA

### POC

https://github.com/Expensify/App/assets/47495619/15c3c12c-1c9e-4624-8f1c-e85fb26938af

**Comment by joekaufmanexpensify** [comment]
Def seems like a bug. @Ollyws thoughts on the above proposals?

**Comment by joekaufmanexpensify** [comment]
Great, ty!

**Comment by barros001** [comment]
@Ollyws Got it. That PR does look pretty similar to what I tried locally, so yeah, let's avoid that. What about my alternative solution? The idea is to simply reset the IOU every time the screen is first loaded. It's similar to what `startMoneyRequest` does, but because the route is being loaded from a link, we don't have a chance to reset it before it loads. So instead we reset it on the component level, when the screen loads using, for example:

```js
    useEffect(() => {
        resetMoneyRequestInfo();
    }, []);
```

Also, we would do it on `MoneyRequestSelectorPage.js` instead of `NewRequestAmountPage.js` as `NewRequestAmountPage`, `NewDistanceRequestPage` and `ReceiptSelector` all depend on `IOU` and not reseting them when on any of the other screens could have unexpected site effects just like this one we're dealing with.

**Comment by maxconnectAbhi** [comment]
@Ollyws  No I am not modifying code inside isEditing block , Its in else condition.

**Comment by peterdbarkerUK** [comment]
@Ollyws - do you think [the bug report here](https://github.com/Expensify/App/issues/30060) is derivative/has the same root cause as this issue?

**Comment by joekaufmanexpensify** [comment]
Ongoing discussion about solutions

**Comment by joekaufmanexpensify** [comment]
@Ollyws curious if you think we're close to being able to select any of these proposals?

**Comment by joekaufmanexpensify** [comment]
Still pending more discussion on proposals

**Comment by barros001** [comment]
@Ollyws what if we reset on mount, but only if url type parameter does not match the current IOU type in Onyx? Downside is clicking on a link from chat would not reset unless type changes but at least it will behave better (this is current behavior anyways, so still a net positive).

Unless there’s a way to intercept the link click, determine that it’s opening either a send or request screen and only then reset (I’ll try to look into this option)

**Comment by barros001** [comment]
## Proposal
[Updated](https://github.com/Expensify/App/issues/29972#issuecomment-1771197062)

**Comment by joekaufmanexpensify** [comment]
Pending more discussion on proposals

**Comment by joekaufmanexpensify** [comment]
@Ollyws are we close to selecting a proposal here?

**Comment by barros001** [comment]
@Ollyws thanks for reviewing my proposal. Looks like [Upwork job failed to post](https://github.com/Expensify/App/issues/29972#issuecomment-1770973916). I'll get a PR ready for review once the job is posted.

**Comment by joekaufmanexpensify** [comment]
@barros001 Hi! As a heads up, your proposal has not been officially selected until you've been assigned to the issue. Right now, we're waiting for @luacmartins to do a final review here. Once you're assigned, you'll be invited to the Upwork job. 

LMK if you have any questions. Thanks!

**Comment by luacmartins** [comment]
Agree that @barros001 proposal looks good. Assigning @barros001 to the issue.

**Comment by barros001** [comment]
I found the Upwork job by searching Upwork (it was created but never linked here). Upwork decided that it needs to verify my account before I can apply for the job (it's my understanding the first time I have to apply, then after that I'd be invited). I started the process but might take a day or two.

**Comment by joekaufmanexpensify** [comment]
All good, and sounds good!

**Comment by barros001** [comment]
Well, it was a lot faster than I thought it would be, my account was verified already and I applied for the job via Upwork.

**Comment by joekaufmanexpensify** [comment]
@barros001 offer sent for $500! Please let us know once the PR is ready for review.

**Comment by barros001** [comment]
[PR](https://github.com/Expensify/App/pull/31009) is ready for review. Just pending attaching video for Android native, will do as soon as I can get it to run on my dev machine.

**Comment by luacmartins** [comment]
I'll put this issue on hold for https://github.com/Expensify/App/pull/28618, since that refactor should fix this.

**Comment by joekaufmanexpensify** [comment]
Yep, I agree! We'll wait and see if that PR fixes this issue. If it does, we'll still issue payment and close this out. If not, we'll proceed with the PR.

**Comment by joekaufmanexpensify** [comment]
Held

**Comment by joekaufmanexpensify** [comment]
Other PR was deployed to staging today. Going to retest there.

**Comment by joekaufmanexpensify** [comment]
Restested, and now when I go to `https://staging.new.expensify.com/request/new` I am redirected to a generic error page. 

<img width="1511" alt="image" src="https://github.com/Expensify/App/assets/54866469/bd44d5e9-51ec-46c8-801f-6ac2a3d40f34">

I think this is because the request refactor changed the format of request URLs. @luacmartins thoughts on fixing this? 

Perhaps that could be a good pivot for this issue. And instead of going to the above page (where there is no option to leave), we could go to `https://staging.new.expensify.com/not-found`, which let you go back to the home page. 

<img width="1511" alt="image" src="https://github.com/Expensify/App/assets/54866469/117aad32-5c18-4b4d-8baa-354f5bb9bd72">

**Comment by barros001** [comment]
Problem here is the following:

https://github.com/Expensify/App/blob/2d8523ae3fd8a5b692700a91efb90adac7f1175f/src/ROUTES.ts#L247-L298

https://github.com/Expensify/App/blob/2d8523ae3fd8a5b692700a91efb90adac7f1175f/src/libs/Navigation/linkingConfig.ts#L393-L410

https://github.com/Expensify/App/blob/2d8523ae3fd8a5b692700a91efb90adac7f1175f/src/pages/iou/MoneyRequestSelectorPage.js#L123-L139

The `/request/new` route has been replaced but it's still there because money request routes will catch `ANYTHING/new` and try to render `MoneyRequestSelectorPage`. If `iouType` is invalid, it will render a not-found screen, but `request` IS valid. The app will continue rendering the old request money screen and it will fail because `CONST.TAB.DISTANCE`, `CONST.TAB.MANUAL`, and `CONST.TAB.SCAN` no longer exist. The same will happen to `split` (not sure if split have been migrated or not).

To me looks like we're in this transition phase where this route/screen has to exist because `send` still uses it. Do we plan to transition `send` as well and eventually get rid of the old screens? If that's the case I suggest we do the following:

Update the should show check to this:

```js
shouldShow={!IOUUtils.isValidMoneyRequestType(iouType) || !isAllowedToCreateRequest || [CONST.IOU.TYPE.REQUEST, CONST.IOU.TYPE.SPLIT].includes(iouType)}
```

And get rid of this `if` block:
https://github.com/Expensify/App/blob/2d8523ae3fd8a5b692700a91efb90adac7f1175f/src/pages/iou/MoneyRequestSelectorPage.js#L111-L141

This will make it so `request/new` behaves the same way as `invalidioutype/new` until everything is migrated away and this routes/screens are completely removed.

**Comment by joekaufmanexpensify** [comment]
Ah, actually I think we just fixed that [here](https://github.com/Expensify/App/issues/32911).

**Comment by joekaufmanexpensify** [comment]
Chatted with @luacmartins and there's nothing else we need to fix here. I just need to pay $500 to each of @barros001 and @Ollyws and then close this out. Will take care of that tomorrow!

**Comment by joekaufmanexpensify** [comment]
@barros001 $500 sent and contract ended!

**Comment by joekaufmanexpensify** [comment]
Upwork job auto closed, so opened a [new one](https://www.upwork.com/jobs/~015f82f561931ee88c) to pay @Ollyws

**Comment by joekaufmanexpensify** [comment]
@Ollyws offer sent for $500!

**Comment by joekaufmanexpensify** [comment]
Paid! Closing as this is all set.
